File: codeforces/1800~1899/1856/e.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

constexpr int MAX_N = 1e6;

void solve() {
  int n = 0;
  cin >> n;
  vector<int> parent(n+1, -1);
  for (int i = 2; i <= n; ++i) cin >> parent[i];

  vector<vector<int>> adj(n+1);
  for (int i = 2; i <= n; ++i) {
    adj[parent[i]].push_back(i);
  }

  int ans = 0;
  function<int(int,int)> dfs = [&](int node, int p) -> int {
    vector<int> counts;
    for (int next_node : adj[node]) if (next_node != p) {
      counts.push_back(dfs(next_node, node));
    }
    // cout << "children size: ";
    // for (int c : counts) cout << c << " ";
    // cout << endl;
    int total = accumulate(counts.begin(), counts.end(), 0);
    vector<bool> dp(total+1);
    dp[0] = true;
    int max_mul = 0;

    for (int i = 0; i < counts.size(); ++i) {
      vector<bool> next_dp = dp;
      for (int a = 0; a+counts[i] <= total; ++a) {
        if (dp[a]) {
          next_dp[a+counts[i]] = true;
          int x = a + counts[i];
          int y = total - x;
          max_mul = max(max_mul, x * y);
        }
      }
      dp = move(next_dp);
    }

    // cout << "max mul=" << max_mul << endl;
    ans += max_mul;
    return total + 1;
  };

  dfs(1, -1);

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1800~1899/1856/e.rs
use std::io;
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::max;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn dfs(adj: &Vec<Vec<usize>>, node: usize, parent: usize, ans: &mut i64) -> i32 {
    let mut sizes: Vec<i32> = Vec::new();
    for next_node in &adj[node] {
        if *next_node == parent {
            continue;
        }

        sizes.push(dfs(&adj, *next_node, node, ans));
    }

    let total = sizes.iter().sum::<i32>() as usize;
    let mut dp: Vec<bool> = vec![false; total+1];
    dp[0] = true;
    let mut max_mul = 0;
    for i in 0..sizes.len() {
        for a in (0..=total).rev() {
            let a2 = (a as i32) - sizes[i];
            if a2 < 0 {
                continue;
            }

            if dp[a2 as usize] {
                dp[a] = true;

                let x = a as i64;
                let y = (total as i64) - x;
                max_mul = max(max_mul, x * y);
            }
        }
    }

    *ans += max_mul;

    (total as i32) + 1
}

fn solve() {
    let n = read_int::<usize>();
    let parents = read_vec::<usize>();

    let mut adj: Vec<Vec<usize>> = vec![Vec::new(); n+1];
    for (i, p) in parents.iter().enumerate() {
        adj[*p].push(i+2);
    }

    let mut ans: i64 = 0;
    dfs(&adj, 1, 0, &mut ans);

    println!("{}", ans);
}

fn main() {
    solve();
}
